extra/modules/fiftyone-devicedetection/src/main/java/org/prebid/server/hooks/modules/fiftyone/devicedetection/v1/core/PipelineBuilderBuilder.java

```diff
@@ -0,0 +1,155 @@
+package org.prebid.server.hooks.modules.fiftyone.devicedetection.v1.core;
+
+import fiftyone.devicedetection.DeviceDetectionOnPremisePipelineBuilder;
+import fiftyone.devicedetection.DeviceDetectionPipelineBuilder;
+import fiftyone.pipeline.core.flowelements.PipelineBuilderBase;
+import fiftyone.pipeline.engines.Constants;
+import fiftyone.pipeline.engines.services.DataUpdateServiceDefault;
+import org.apache.commons.lang3.BooleanUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.prebid.server.hooks.modules.fiftyone.devicedetection.model.config.DataFile;
+import org.prebid.server.hooks.modules.fiftyone.devicedetection.model.config.DataFileUpdate;
+import org.prebid.server.hooks.modules.fiftyone.devicedetection.model.config.ModuleConfig;
+import org.prebid.server.hooks.modules.fiftyone.devicedetection.model.config.PerformanceConfig;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class PipelineBuilderBuilder {
+    private static final Collection<String> PROPERTIES_USED = List.of(
+            "devicetype",
+            "hardwarevendor",
+            "hardwaremodel",
+            "hardwarename",
+            "platformname",
+            "platformversion",
+            "screenpixelsheight",
+            "screenpixelswidth",
+            "screeninchesheight",
+            "pixelratio",
+
+            "BrowserName",
+            "BrowserVersion",
+            "IsCrawler",
+
+            "BrowserVendor",
+            "PlatformVendor",
+            "Javascript",
+            "GeoLocation",
+            "HardwareModelVariants");
+
+    private DeviceDetectionOnPremisePipelineBuilder premadeBuilder = null;
+
+    public PipelineBuilderBuilder withPremadeBuilder(DeviceDetectionOnPremisePipelineBuilder premadeBuilder) {
+        this.premadeBuilder = premadeBuilder;
+        return this;
+    }
```

Comment:
as I can judge premadeBuilder is used only for testing purpose, this is the bad practice

please compare the builder is expected OR I suggest returning a Pipeline object on calling the `build`, in that way you can check whether the Pipeline has the state you expected after building

Reply:
I've got the following explanation from @drasmart why this is implemented this way (if I understood and paraphrased it correctly):

* [[PipelineDefault](https://github.com/51Degrees/pipeline-java/blob/main/pipeline.core/src/main/java/fiftyone/pipeline/core/flowelements/PipelineDefault.java)] class is inaccessible.
  * The state of this class is not a part of the `Pipeline` interface contract.
  * The `PipelineBuilder` classes (of library, not module) are required to return a `Pipeline` that behaves in a certain way, and it is the library's internal concern, not modules - so the state is tested separately in the library.
* `PipelineBuilderBuilder` is the boundary class (between the module and library code, or one that is wrapping library code).
  * It's responsibility is invoking side-effectful foreign (library's) APIs.
  * Counting invocations of the APIs of the passed mock object is done to test that the behavior actually supports the contract.
  
Please let us know if this is an ok approach to test it this way, or should we seek some other way to implement this boundary/wrapper.  Thanks.

Reply:
## The primary option:

PipelineDefault is using internally and it's fine, but the PipelineBuilder builds that object, and further in the code the Pipeline interface represents it. And that's totally fine. From my perspective you don't need to test the building itself, you can check the output whether the result Pipeline behaves in the way you expect from it. 

I don't understand why you call "checking whether the Pipeline is built correctly" an internal concern if you build it inside the module and use the result of this building further in the code.

What is you're trying to cover is only the building steps, that only proves that the builder calls setters, but the output result and it's validness is not clear. 

## The alternative (I don't like but I don't understand why the primary one is not sufficient)

```java
public class PipelineBuilder {
    
    private final ModuleConfig moduleConfig;

    public PipelineBuilder(ModuleConfig moduleConfig) {
        this.moduleConfig = moduleConfig;
    }

    public Pipeline build(DeviceDetectionPipelineBuilder premadeBuilder) throws Exception {
        final DataFile dataFile = moduleConfig.getDataFile();
        
        final Boolean shouldMakeDataCopy = dataFile.getMakeTempCopy();
        final DeviceDetectionOnPremisePipelineBuilder builder = premadeBuilder.useOnPremise(
                dataFile.getPath(), 
                BooleanUtils.isTrue(shouldMakeDataCopy));
        
        applyUpdateOptions(builder, dataFile.getUpdate());
        applyPerformanceOptions(builder, moduleConfig.getPerformance());
        PROPERTIES_USED.forEach(builder::setProperty);
        return builder.build();
    }
    
    ...

}
```
and the hook
```java
    public FiftyOneDeviceDetectionRawAuctionRequestHook(ModuleConfig moduleConfig) throws Exception {
        this.moduleConfig = Objects.requireNonNull(moduleConfig);
        final Pipeline pipeline = new PipelineBuilder(moduleConfig).build(new DeviceDetectionPipelineBuilder());
        this.deviceEnricher = new DeviceEnricher(pipeline);
    }
```